cpp/demo/poisson_matrix_free/main.cpp

```diff
@@ -0,0 +1,215 @@
+    const U alpha = rnorm / la::inner_product(p, y);
+
+    // Update x (x <- x + alpha*p)
+    axpy(x, alpha, p, x);
+
+    // Update r (r <- r - alpha*y)
+    axpy(r, -alpha, y, r);
+
+    // Update residual norm
+    // Note: we use U for beta to support float, double, etc. U can be
+    // complex, even though the value will always be real
+    const double rnorm_new = r.squared_norm();
+    const U beta = rnorm_new / rnorm;
+    rnorm = rnorm_new;
+
+    if (rnorm / rnorm0 < rtol2)
+      break;
+
+    // Update p (p <- beta*p + r)
+    axpy(p, beta, p, r);
+  }
+
+  return k;
+}
+} // namespace linalg
+
+int main(int argc, char* argv[])
+{
+  common::subsystem::init_logging(argc, argv);
+  common::subsystem::init_mpi(argc, argv);
+
+  {
+    using T = PetscScalar;
+
+    MPI_Comm comm = MPI_COMM_WORLD;
+
+    // Create mesh and function space
+    auto mesh = std::make_shared<mesh::Mesh>(mesh::create_rectangle(
+        comm, {{{0.0, 0.0}, {1.0, 1.0}}}, {10, 10}, mesh::CellType::triangle,
+        mesh::GhostMode::none));
+
+    auto V = std::make_shared<fem::FunctionSpace>(
+        fem::create_functionspace(functionspace_form_poisson_M, "ui", mesh));
+
+    // Prepare and set Constants for the bilinear form
+    auto f = std::make_shared<fem::Constant<T>>(-6.0);
+
+    // Define variational forms
+    auto L = std::make_shared<fem::Form<T>>(
+        fem::create_form<T>(*form_poisson_L, {V}, {}, {{"f", f}}, {}));
+
+    // Action of the bilinear form "a" on a function ui
+    auto ui = std::make_shared<fem::Function<T>>(V);
+    auto M = std::make_shared<fem::Form<T>>(
+        fem::create_form<T>(*form_poisson_M, {V}, {{"ui", ui}}, {{}}, {}));
+
+    // Define boundary condition
+    auto u_D = std::make_shared<fem::Function<T>>(V);
+    u_D->interpolate(
+        [](auto&& x) {
+          return 1 + xt::square(xt::row(x, 0)) + 2 * xt::square(xt::row(x, 1));
+        });
+
+    std::vector<std::int32_t> facets = mesh::exterior_facet_indices(*mesh);
+    std::vector<std::int32_t> bdofs
+        = fem::locate_dofs_topological({*V}, 1, facets);
+    auto bc = std::make_shared<const fem::DirichletBC<T>>(u_D, bdofs);
+
+    // Assemble RHS vector
+    la::Vector<T> b(V->dofmap()->index_map, V->dofmap()->index_map_bs());
+    fem::assemble_vector(b.mutable_array(), *L);
+
+    // Apply lifting to account for Dirichlet boundary condition
+    // b <- b - A * x_bc
+    fem::set_bc(ui->x()->mutable_array(), {bc}, -1.0);
+    dolfinx::fem::assemble_vector(b.mutable_array(), *M);
+
+    // Communicate ghost values
+    b.scatter_rev(common::IndexMap::Mode::add);
+
+    // Set BC dofs to zero (effectively zeroes columns of A)
+    fem::set_bc(b.mutable_array(), {bc}, 0.0);
+
+    // Create function for computing the action of A on x (y = Ax)
+    std::function<void(la::Vector<T>&, la::Vector<T>&)> action
+        = [&](la::Vector<T>& x, la::Vector<T>& y)
+    {
+      // Update ghost values and zero y
+      x.scatter_fwd();
+      y.set(0);
```

Comment:
use floating point literal 0.0

Reply:
Not sure it's required here - what's there is equivalent to `T(0)` I think. Unlikely, but we we could have a vector of `int` :)